tests/python/pants_test/pants_run_integration_test.py

```diff
@@ -32,17 +32,24 @@
 
 
 class PantsResult(datatype([
+    # NB: May be either a string list or string (in the case of `shell=True`).
     'command',
     ('returncode', int),
-    'stdout_data',
-    'stderr_data',
-    'workdir',
+    ('stdout_data', string_type),
+    ('stderr_data', string_type),
+    ('workdir', string_type),
     ('pid', Exactly(*IntegerForPid)),
 ])):
   pass
 
 
-class PantsJoinHandle(datatype(['command', 'process', 'workdir'])):
+class PantsJoinHandle(datatype([
+    # NB: May be either a string list or string (in the case of `shell=True`).
```

Comment:
This is something we may be able to represent with `TypeConstraint`s without too much overhead (and really what I'd like to do is to allow `enum` to support wrapping a value in addition to a tag). I'll make an issue.